### PR TITLE
make DICOM read work for fractional rescale

### PR DESCRIPTION
**Pull request by jcupitt**
The DICOM reader was multiplying pixel values by `rescale_slope`, which it
represented as `ssize_t`. Many DICOM files use double values for
`rescale_slope`, so for `rescale_slope` < 1, the reader was multiplying by
zero and producing blank images.

This patch changes `rescale_slope` and `rescale_intercept` to be double, so
dcm read now works with fractional values of rescale.
This patch should be applied to the 7.x branch as well.

This patch incorporates a couple of other small changes too:

* Previous versions of ImageMagick did not apply the rescale
  values, so for compatibility, application of them now defaults to
  off. Use `-define dcm:rescale` to enable rescale processing.

* `window_center` and `window_width` should also be double-valued, from
  the spec.

* The `window_center` and `window_width` fields were always applied before,
  but from the DICOM spec, they should be only be applied to the rescaled
  values.

  The patch makes width and center processing only happen if `dcm:rescale`
  is on.

* Previously, the offset for signed data was only being applied if
  windowing was disabled. It is now always applied.

Here's a sample file that shows the problem:

http://www.rollthepotato.net/~john/MRIm5.dcm

You're welcome to use this image in a test suite. Before this patch:

$ identify -verbose MRIm5.dcm
Image: MRIm5.dcm
Format: DCM (Digital Imaging and Communications in Medicine image)
Gray:
min: 32769 (0.500023)
max: 32769 (0.500023)

After:

$ identify -verbose MRIm5.dcm
Image: MRIm5.dcm
Format: DCM (Digital Imaging and Communications in Medicine image)
Gray:
min: 0 (0)
max: 65535 (1)

Forum thread:

https://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=31887&p=145852

Related PR:

https://github.com/ImageMagick/ImageMagick/pull/477

### CODE REVIEW

**Comment by jcupitt** [comment]
If/when this patch is accepted, I'll make another PR which adds `dcm:window` so that the window center and width can be overridden from the command-line.

File: coders/dcm.c [hunk 1]
```diff
@@ -4076,6 +4095,10 @@ static Image *ReadDCMImage(const ImageInfo *image_info,ExceptionInfo *exception)
             if (LocaleCompare(option,"reset") == 0)
               info.window_width=0;
           }
+        option=GetImageOption(image_info,"dcm:rescale");
+        if (option != (char *) NULL)
+          info.rescale=(int) ParseCommandOption(MagickBooleanOptions,
```
Comment (dlemstra) [inline coders/dcm.c]: You could use `IsStringTrue` here.

File: coders/dcm.c [hunk 2]
```diff
@@ -3080,20 +3094,21 @@ static Image *ReadDCMImage(const ImageInfo *image_info,ExceptionInfo *exception)
...
-  info.rescale_slope=1;
+  info.max_value=255UL;
   info.samples_per_pixel=1;
-  info.scale=(Quantum *) NULL;
   info.signed_data=(~0UL);
   info.significant_bits=0;
-  info.window_center=0;
-  info.window_width=0;
-  info.polarity=MagickFalse;
+  info.rescale=0;
```
Comment (dlemstra) [inline coders/dcm.c]: This should be MagickFalse instead.

File: coders/dcm.c [hunk 3]
```diff
@@ -2877,32 +2880,43 @@ static MagickBooleanType ReadDCMPixels(Image *image,DCMInfo *info,
...
-              else
-                if ((ssize_t) index > window_max)
-                  index=(int) info->max_value;
-                else
-                  index=(int) (info->max_value*(((index-info->window_center-
-                    0.5)/(info->window_width-1))+0.5));
+              index=pixel_value;
             }
+          if (info->signed_data == 1)
+            index-=32767;
```
Comment (dlemstra) [inline coders/dcm.c]: The old code did not do this when the `windows_size` was not 0. Is this change intentional?

**Comment by dlemstra** [comment]
We would also welcome a patch for our IM6 branch 👍  (*hint*)

File: coders/dcm.c (hunk 3 above)
Comment (jcupitt) [inline coders/dcm.c]: Yes, it seemed correct to me. If the output is signed, you must offset the origin, whether or not you interpret the window. I have not test files I could check with, though.

File: coders/dcm.c (hunk 1 above)
Comment (jcupitt) [inline coders/dcm.c]: Oh nice. I copy-pasted that code from the webp loader, perhaps that should change too.

**Comment by jcupitt** [comment]
This patch is for the IM-6 branch, I agree a patch to master would be nice too. Should I make another PR once this is done?

**Comment by jcupitt** [comment]
OK, pushed the changes you proposed. Thanks for the review!

**Comment by dlemstra** [comment]
Feel free to wait with the IM7 patch until we have applied the `dcm:window` changes in IM6.

**Comment by jcupitt** [comment]
\o/ thanks! I'll make another PR for the window stuff.

**Comment by jcupitt** [comment]
`dcm:window` PR is here

https://github.com/ImageMagick/ImageMagick/pull/484
